examples/java8/src/main/java/org/apache/beam/examples/complete/game/StatefulTeamScore.java

```diff
@@ -0,0 +1,224 @@
+                options.as(GcpOptions.class).getProject(),
+                options.getDataset(),
+                options.getLeaderBoardTableName() + "_team_leader",
+                configureCompleteWindowedTableWrite()));
+
+    // Run the pipeline and wait for the pipeline to finish; capture cancellation requests from the
+    // command line.
+    PipelineResult result = pipeline.run();
+    exampleUtils.waitToFinish(result);
+  }
+
+  /**
+   * Tracks each team's score separately in a single state cell and outputs the score every time it
+   * passes a new multiple of a threshold.
+   *
+   * <p>We use stateful {@link DoFn} because:
+   * <ul>
+   *   <li>State is key-partitioned. Therefore, the score is calculated per team.</li>
+   *   <li>Stateful {@link DoFn} can determine when to output based on the state. This allows only
+   *       outputting when a team's score passes a given threshold.</li>
+   * </ul></p>
+   */
+  @VisibleForTesting
+  public static class UpdateTeamScoreFn
+      extends DoFn<KV<String, GameActionInfo>, KV<String, Integer>> {
+
+    private static final String TOTAL_SCORE = "totalScore";
+    private final int thresholdScore;
+
+    public UpdateTeamScoreFn(int thresholdScore) {
+      this.thresholdScore = thresholdScore;
+    }
+
+    /**
+     * Describes the state for storing team score. Let's break down this statement.
+     *
+     * {@link StateSpec} configures the state cell, which is provided by a runner during pipeline
+     * execution.
+     *
+     * {@link org.apache.beam.sdk.transforms.DoFn.StateId} annotation assigns an identifier to the
+     * state, which is used to refer the state in @ProcessElement.
+     *
+     * <p>A {@link ValueState} stores single value per key and per window. Because our pipeline is
+     * globally windowed in this example, this {@link ValueState} is just key partitioned, with one
+     * score per team. Any other class that extend {@link org.apache.beam.sdk.state.State} can be
+     * used.</p>
+     *
+     * <p>In order to store the value, the state must be encoded. Therefore, we provide a coder, in
+     * this case the {@link VarIntCoder}. If the coder is not provided as StateSpecs.value(), Beam's
+     * coder inference will try to provide a coder automatically.</p>
+     */
+    @StateId(TOTAL_SCORE)
+    private final StateSpec<ValueState<Integer>> totalScoreSpec =
+        StateSpecs.value(VarIntCoder.of());
+
+    /**
+     * To use a state cell, annotate a parameter with @StateId that matches the state declaration.
+     * The type of the parameter should match the StateSpec type.
+     */
+    @ProcessElement
+    public void processElement(
+        ProcessContext c,
+        @StateId(TOTAL_SCORE) ValueState<Integer> totalScore) {
+      String teamName = c.element().getKey();
+      GameActionInfo gInfo = c.element().getValue();
+
+      // State is read and modified by using read(), write() methods.
```

Comment:
I don't think this comment is useful. It may be useful to include a note among the lines of "ValueState cells do not contain a default value, so use an explicit one."

Reply:
Changed the comment to "ValueState cells do not contain a default value. If the state is possibly not written, make sure to check for null on read."